port/unix/omrsysinfo.c

```diff
@@ -3071,3 +3161,360 @@ omrsysinfo_os_kernel_info(struct OMRPortLibrary *portLibrary, struct OMROSKernel
+{
+	OMRCgroupEntry *temp = cgEntryList;
+	char *cgName = NULL;
+
+	if (NULL == temp) {
+		goto _end;
+	}
+
+	do {
+		if (!strcmp(temp->subsystem, subsystem)) {
+			cgName = temp->cgroup;
+			break;
+		}
+		temp = temp->next;
+	} while (temp != cgEntryList);
+
+_end:
+	return cgName;
+}
+
+/**
+ * Allocates a new OMRCgroupEntry structure and adds it to the circular linked list pointed by *cgEntryList.
+ * On success, *cgEntryList points to the new OMRCgroupEntry structure.
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] cgEntryList pointer to OMRCgroupEntry* which points to circular linked list
+ * @param[in] hierId hierarchy Id of the cgroup
+ * @param[in] subsystem name of cgroup subsystem/resource controller
+ * @param[in] cgroupName name of the cgroup
+ *
+ * @return 0 on success, negative error code on failure
+ */
+static int32_t
+addCgroupEntry(struct OMRPortLibrary *portLibrary, OMRCgroupEntry **cgEntryList, int32_t hierId, const char *subsystem, const char *cgroupName)
+{
+	OMRCgroupEntry *cgEntry = NULL;
+	int32_t cgEntrySize = sizeof(OMRCgroupEntry) + strlen(subsystem) + 1 + strlen(cgroupName) + 1;
+	int32_t rc = 0;
+
+	cgEntry = portLibrary->mem_allocate_memory(portLibrary, cgEntrySize, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+	if (NULL == cgEntry) {
+		rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_MEMORY_ALLOC_FAILED, "memory allocation for cgroup entry failed");
+		goto _end;
+	}
+	cgEntry->hierarchyId = hierId;
+	cgEntry->subsystem = (char *)(cgEntry + 1);
+	strcpy(cgEntry->subsystem, subsystem);
+	cgEntry->cgroup = cgEntry->subsystem + strlen(subsystem) + 1;
+	strcpy(cgEntry->cgroup, cgroupName);
+
+	if (NULL == *cgEntryList) {
+		*cgEntryList = cgEntry;
+		cgEntry->next = cgEntry; /* create a circular list */
+	} else {
+		OMRCgroupEntry *first = *cgEntryList;
+		*cgEntryList = cgEntry;
+		cgEntry->next = first->next;
+		first->next = cgEntry;
+	}
+
+_end:
+	return rc;
+}
+
+/**
+ * Reads /proc/<pid>/cgroup to get information about the cgroups to which the given
+ * process belongs.
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] pid process id
+ * @param[out] cgroupEntryList pointer to OMRCgroupEntry *. On successful return, *cgroupEntry
+ * points to a circular linked list. Each element of the list is populated based on the contents 
+ * of /proc/<pid>/cgroup file. 
+ *
+ * returns 0 on success, negative code on error
+ */
+static int32_t
+readCgroupFile(struct OMRPortLibrary *portLibrary, int pid, OMRCgroupEntry **cgroupEntryList)
+{
+	char cgroup[PATH_MAX];
+	/* This array should be large enough to read names of all subsystems. 1024 should be enough. */
+	char subsystems[1024];
+	FILE *cgroupFile = NULL;
+	OMRCgroupEntry *cgEntryList = NULL;
+	int32_t rc = 0;
+
+	Assert_PRT_true(NULL != cgroupEntryList);
+	
+	portLibrary->str_printf(portLibrary, cgroup, sizeof(cgroup), "/proc/%d/cgroup", pid);
```

Comment:
I noticed that but didn't bother to change because, as you said, "/proc/%d/cgroup" is unlikely to be more than PATH_MAX.
I am wondering what kind of assert are you suggesting. We don't know the actual length of the string - "/proc/%d/cgroup".